lib/puppet/provider/nssdb/certutil.rb

```diff
@@ -0,0 +1,50 @@
+require 'fileutils'
+
+Puppet::Type.type(:nssdb).provide(:certutil) do
+  commands :certutil => 'certutil'
+
+  def create
+    create_nssdb_dir
+    generate_nssdb
+  end
+
+  def destroy
+    destroy_nssdb
+  end
+
+  def exists?
+    nssdb_exists?
+  end
+
+  private
+
+  def create_nssdb_dir
+    FileUtils.mkdir_p(resource[:nssdb_dir])
+  end
+
+  def generate_nssdb
+    certutil(
+      '-N',
+      '-d', resource[:nssdb_dir],
+      '-f', resource[:password_file]
+    )
+  rescue Puppet::ExecutionFailure => e
+    Puppet.err("Failed to generate new NSS database at #{resource[:nssdb_dir]} with password file #{resource[:password_file]}: #{e}")
+    return nil
+  end
+
+  def destroy_nssdb
+    FileUtils.rm_rf(resource[:nssdb_dir], secure: true)
+  end
+
+  def nssdb_exists?
+    certutil(
+      '-L',
+      '-d', resource[:nssdb_dir],
+      '-f', resource[:password_file]
+    )
+  rescue Puppet::ExecutionFailure
+    Puppet.debug("Did not find existing NSS database at #{resource[:nssdb_dir]}")
```

Comment:
Is an execution failure always this? Can't there be some other error? Like an invalid password in the password file.

Reply:
Great point.

I'll make the change to

```ruby
  rescue Puppet::ExecutionFailure => e
    Puppet.debug("Unable to verify NSS database at #{resource[:nssdb_dir]} with password file #{resource[:password_file]}: #{e}")
```

Open to alternative wording if you have strong feelings about it.

BTW, I think we are OK with treating incorrect password file contents / location as -> NSS database does not exist and should be (re)-created. That should subsequently trigger a refresh of the `nssdb_certificate` resources, which should still add the same certificate content as before to the new NSS db.